esp-wifi/src/ble/controller/mod.rs

```diff
@@ -176,4 +183,43 @@ pub mod asynch {
             }
         }
     }
+
+    impl<'d> Transport for BleConnector<'d> {
+        /// Read a complete HCI packet into the rx buffer
+        async fn read<'a>(
+            &self,
+            rx: &'a mut [u8],
+        ) -> Result<ControllerToHostPacket<'a>, Self::Error> {
+            if !have_hci_read_data() {
+                HciReadyEventFuture.await;
+            }
+
+            let mut total = 0;
+            for b in rx.iter_mut() {
+                let mut buffer = [0u8];
+                let len = read_hci(&mut buffer);
+
+                if len == 1 {
+                    *b = buffer[0];
```

Comment:
```suggestion
                let len = read_hci(core::slice::from_mut(b));

                if len == 1 {
```

To be honest I'm not entirely sure why this (and the existing embedded-io implementation) reads a single byte at a time but it makes my eye twitch.

Looking at the implementation, it should be safe to (try) read multiple bytes at a time. Not a problem for this PR though.

Reply:
> To be honest I'm not entirely sure why this (and the existing embedded-io implementation) reads a single byte at a time

Let's say the original code was written as a POC for another MCU 😉 and since it works it was never really touched since then - but for sure there are things which could get improved

Reply:
The integration should be more correct now as well using read_next rather than the 'byte-oriented' read_hci interface.